### PR TITLE
Added MiniMessage support to built-in motd

### PR DESCRIPTION
**Pull request by 4drian3d**
- MiniMessage support has been added to build-in motd replacing legacy color codes and json usage
- Bump config-version to 2.6

### CODE REVIEW

File: proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java]: Nit: needs to be `<mm><dark_aqua>A Velocity Server`

**Comment by astei** [comment]
I'm happy with this but would like some input from others before I merge.

**Review by KoxSosen — APPROVED** [review]
I'm not a contributor, but this seems like a great change.

**Comment by kezz** [comment]
A prefix is really weird - is there any reason not to migrate the config?

**Comment by kyngs** [comment]
> A prefix is really weird - is there any reason not to migrate the config?

That would mean peoples motd's breaking on a minor update. Not having the prefix would be a good idea on velocity 4.x IMO.

**Comment by 4drian3d** [comment]
> A prefix is really weird - is there any reason not to migrate the config?

I thought it would be the best option, to migrate the formatting of that option in a "patch" version like 3.1.2 would be strange

**Comment by kezz** [comment]
> That would mean peoples motd's breaking on a minor update. Not having the prefix would be a good idea on velocity 4.x IMO.

Do you know what migration means? The point of migration is that nothing breaks.

> to migrate the formatting of that option in a "patch" version like 3.1.2 would be strange

Then let's bump the minor version? I assume Velocity, being a modern proxy, would want to migrate away from legacy formatting eventually, so why not do it in one step?

**Comment by kyngs** [comment]
> > That would mean peoples motd's breaking on a minor update. Not having the prefix would be a good idea on velocity 4.x IMO.
> 
> Do you know what migration means? The point of migration is that nothing breaks.
> 

I'm sorry, I understood it as if it were to just switch to mm. How do you want to migrate such a thing?

**Comment by Astralchroma** [comment]
I made a similar pull request to Paper, and after discussion there the course of action was to forcibly migrate the config messages instead of maintain compatibility with both, would it not be best to be consistent and forcibly migrate the format with Velocity?

**Comment by 4drian3d** [comment]
> > That would mean peoples motd's breaking on a minor update. Not having the prefix would be a good idea on velocity 4.x IMO.
> 
> Do you know what migration means? The point of migration is that nothing breaks.
> 
> > to migrate the formatting of that option in a "patch" version like 3.1.2 would be strange
> 
> Then let's bump the minor version? I assume Velocity, being a modern proxy, would want to migrate away from legacy formatting eventually, so why not do it in one step?

Yes, the ideal would be to migrate at once from all legacy color code uses to MiniMessage, but being a minor version to be released in proximity, I thought better to use a prefix, but if required, I could try to migrate to MiniMessage in that option now

**Comment by kezz** [comment]
> How do you want to migrate such a thing?

Velocity has a versioned config, there's a million and one ways you could easily setup migrations.

**Comment by 4drian3d** [comment]
> > How do you want to migrate such a thing?
> 
> Velocity has a versioned config, there's a million and one ways you could easily setup migrations.

Honestly, I never paid attention to this configuration option, I just remembered it existed, now I am making the migration

**Comment by Astralchroma** [comment]
MiniMessage can convert components to MiniMessage so it's not hard to convert it.

**Comment by 4drian3d** [comment]
migration1
The migration and the motd format are working correctly, it should be ready for another review

**Comment by astei** [comment]
I'm happy with this but am going to hold this off until we get a new minor version out.

**Comment by wiktor-jozwiak** [comment]
will this be implemented? (About when this version of Velocity will come out)

**Comment by 4drian3d** [comment]
> I'm happy with this but am going to hold this off until we get a new minor version out.

Considering that pull request #692 was accepted and that very soon pull request #690 will be accepted, I believe that the next version will no longer be considered as a "patch" and this pull request could be accepted directly.

Also, it would be nice if the first version that includes MiniMessage already has its use implemented in all the places where the legacy format was used (where possible) which is the only place that would be missing

File: proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java
Comment (kezz) [inline proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java]: Might it be a good time to ditch JSON support too in this tbh. There are valid JSON text components that don't start with `{` and MiniMessage can represent everything that JSON can so there's no need to still have it.

Comment (4drian3d) [reply proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java]: According to what you say, I just checked and any slightly complex JSON formatting in this option was unusable since the [same check](https://github.com/PaperMC/Velocity/blob/dev/3.0.0/proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java#L276) was used to parse JSON 😅

I just added migration support for the JSON format, but I had to add 2 additional `.replace("\\", "")` because otherwise the Gson serializer would add that character, I don't know if it's related to the fix https://github.com/KyoriPowered/adventure/pull/789 but it doesn't work

File: proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java
Comment (kezz) [inline proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java:277]: There's a deserializeOrNull method you can use just to save the if check here

File: proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java
Comment (kezz) [inline proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java]: imo it's only worth doing exactly the same check that was done when the component was retrieved originally (just the `{` check), as nothing else would've been displayed as JSON anyway

File: proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java
Comment (4drian3d) [reply proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java:277]: I think it is better to leave it as it is, since the only time the motd of the configuration changes, is when the whole configuration is reloaded, so it would be an unnecessary waste of resources to always call the `MiniMessage#deserialize` or `#deserializeOrNull` method if the motd will always be the same

File: proxy/src/main/resources/default-velocity.toml
Comment (kezz) [inline proxy/src/main/resources/default-velocity.toml]: need to remove the JSON bit here

**Review by astei — APPROVED** [review]

**Pull request merged**
